### PR TITLE
Update SDPA to PagedAttention transformation to support Qwen model

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
Updated SDPA to PagedAttention transformation to support Qwen model case.
Fixed SDPA to PagedAttention  transformation to copy friendly/tensor names correctly.
Added unit tests.

### Tickets:
 - *CVS-157308*

### CODE REVIEW

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp [hunk 1]
```diff
@@ -7,37 +7,85 @@
...
+        // are used not in the sequential order, so we need to use position_ids to get the expected values.
+        auto gather = std::make_shared<v8::Gather>(slice_1, position_ids, axis);
+        gather->set_friendly_name(slice_2->get_friendly_name());
+        gather->validate_and_infer_types();
+
+        // PagedAttention expects the next layout for Q,K,V:
+        // [batch_size_in_tokens, num_kv_heads * head_size]
+        // so here we need to reshape the output tensor to move the seq dim (num tokens) to the batch
+        // num_kv_heads * head_size are already handled in the StateManagementPattern transformation
+        auto head_size = static_cast<int64_t>(rotary_emb_sincos->get_shape()[3]);
```
Comment (itikhono) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp:82]: probably need to check a rank

File: src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp [hunk 2]
```diff
@@ -23,4 +24,21 @@ class ov::pass::PositionIDsReplacer : public ov::pass::MatcherPass {
 public:
     OPENVINO_RTTI("PositionIDsReplacer", "0");
     explicit PositionIDsReplacer(const Output<Node>& position_ids);
-};
\ No newline at end of file
+};
+
+/**
+ * @brief Qwen model has a specific feature in the model structure not to use position_ids input,
+ * this input is detached. The model expects data processing in order.
```
Comment (slyalin) [inline src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp:31]: The model uses implicitly defined `position_ids` based on the past KV cache size.

File: src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp [hunk 3]
```diff
@@ -23,4 +24,21 @@ class ov::pass::PositionIDsReplacer : public ov::pass::MatcherPass {
...
+ *
+ * To use this model in Continuous batching mode, we need to apply position_ids and
+ * use the corresponding rotary_emb_cos/rotary_emb_sin.
+ * For this, we replace
+ *      rotary_emb_cos/rotary_emb_sin -> Slice -> Slice
+ * With
+ *      rotary_emb_cos/rotary_emb_sin -> Slice -> Gather(by position_ids)
```
Comment (slyalin) [inline src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp:38]: ... which enables applying RoPE for each token independently of their order in the input tensor.

Comment (slyalin) [reply src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp:38]: Why do we need to keep the first Slice?

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp [hunk 4]
```diff
@@ -7,37 +7,85 @@
...
+    MATCHER_SCOPE(PositionIDsReplacerQwen);
+
+    auto p_max_context_len = wrap_type<v0::Parameter>();
+    auto p_opt_convert = optional<v0::Convert>(p_max_context_len);
+    auto p_opt_reshape = optional<v1::Reshape>({p_opt_convert, any_input()});
+
+    auto p_rotary_emb_sincos = wrap_type<v0::Constant>();
+    // the rotary_emb_cos/rotary_emb_sin are sliced by the total length [1,..4096,1,128]
+    auto p_slice_1 = wrap_type<v8::Slice>({p_rotary_emb_sincos, any_input(), p_opt_reshape, any_input(), any_input()});
+    auto p_slice_2 = wrap_type<v8::Slice>({p_slice_1, any_input(), any_input(), any_input(), any_input()});
```
Comment (slyalin) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp:59]: Is there any chance to set stricter requirements for `begin` and `end` inputs for these two `Slice`s? We would like to move toward conservative patterns in the transformation, but here we are introducing another speculative extension.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp [hunk 5]
```diff
@@ -6,27 +6,33 @@
...
+    auto kv_past = wrap_type<v6::ReadValue>({any_input()});
+    auto kv_gather = wrap_type<v8::Gather>({kv_past, any_input(), any_input()});
+    auto kv_current = any_input();
+    auto kv_concat = wrap_type<v0::Concat>({kv_gather, kv_current});
+    auto kv_shape = wrap_type<v3::ShapeOf>({kv_concat});
+    auto gather_idx_label = wrap_type<v0::Constant>();
+    auto seq = wrap_type<v8::Gather>({kv_shape, gather_idx_label, any_input()});
 
-    ov::matcher_pass_callback callback = [=](ov::pass::pattern::Matcher& m) {
+    ov::matcher_pass_callback callback = [=](Matcher& m) {
```
Comment (slyalin) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:35]: It is a good simplification, but do we really need to mix it with the main subject of this PR? Does it make sense to have another PR where such simplification is applied to all related transformations?

File: src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+enum QKV : int { Q = 0, K = 1, V = 2 };
+vector<int> MOCK_VALUE = {1};
+
+class Qwen7bChatSDPA {
+public:
+    static std::shared_ptr<Node> gen_embeddings(const std::shared_ptr<Node>& input_ids) {
+        auto view_reshape = makeOP<v1::Reshape>({input_ids, {-1, 0}}, {special_zero_true});
+        auto input_ids_i64 = makeOP<v0::Convert>({view_reshape}, {dest_type_i64});
+
+        auto weights = makeConst(element::u8, {151936, 4096}, MOCK_VALUE);
```
Comment (slyalin) [inline src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp:80]: Why do we need to have such a big size in a test? It is 600MB.
How is this part of the model graph related to proposed changes?

File: src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+        auto pa_aligned = Qwen7bChatPA::align_pa_layout(pa, head_size_2);
+        auto res = makeOP<v0::Result>({pa_aligned});
+
+        model_ref = std::make_shared<ov::Model>(OutputVector{res}, params);
+    }
+    // TODO: align precisions, check the copying of "fuse_names" attr in SDPAToPagedAttention
+    // checking the graph structure and names, other checks are temporarily disabled:
+    comparator.disable(FunctionsComparator::PRECISIONS);
+    disable_rt_info_check();
+}
```
Comment (slyalin) [inline src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp:537]: Let me guess: coding OV models as a C++ code in tests is the main reason why the tests compilation is so time-consuming, right? It is a joke (I hope), but why do we need to code all these parts here as C++ code but not just `read_model` them from IR files located somewhere? I'm not against representing parts of the models in the code, I'm just concerned that you are spending more time preparing this code in comparison to the code that reads from an IR file. You are using a tool to generate a body for each part, but you still need to organize parts as a whole. Or are you also using a tool to do that as well?

File: src/core/src/pass/sdpa_to_paged_attention.cpp [hunk 8]
```diff
@@ -81,9 +81,12 @@ bool ov::pass::SDPAToPagedAttention::run_on_model(const std::shared_ptr<ov::Mode
     OPENVINO_ASSERT(input_ids_node, "The model doesn't contain input_ids or input_embeds input. Aborting.");
 
     input_ids_node->set_partial_shape(PartialShape{-1});
+    auto input_ids_target_inputs = input_ids_node->get_output_target_inputs(0);
     auto unsqueezed_input_ids =
         std::make_shared<v0::Unsqueeze>(input_ids_node, v0::Constant::create(element::i32, Shape{}, {1}));
-    replace_node(input_ids_node, unsqueezed_input_ids);
+    for (const auto& target : input_ids_target_inputs) {
+        target.replace_source_output(unsqueezed_input_ids);
+    }
```
Comment (slyalin) [inline src/core/src/pass/sdpa_to_paged_attention.cpp:89]: Hey, what's going on here? `replace_node` looks to be banned everywhere. Why is in this place? And why is in this PR?

File: src/core/src/pass/sdpa_to_paged_attention.cpp [hunk 9]
```diff
@@ -106,9 +109,12 @@ bool ov::pass::SDPAToPagedAttention::run_on_model(const std::shared_ptr<ov::Mode
         position_ids->set_partial_shape(PartialShape{-1});
         position_ids->validate_and_infer_types();
     }
+    auto position_ids_target_inputs = position_ids->get_output_target_inputs(0);
     auto unsqueezed_position_ids =
         std::make_shared<v0::Unsqueeze>(position_ids, v0::Constant::create(element::i32, Shape{}, {1}));
-    replace_node(position_ids, unsqueezed_position_ids);
+    for (const auto& target : position_ids_target_inputs) {
+        target.replace_source_output(unsqueezed_position_ids);
+    }
```
Comment (slyalin) [inline src/core/src/pass/sdpa_to_paged_attention.cpp:117]: Why was it wrong?

Comment (itikhono) [reply src/core/src/pass/sdpa_to_paged_attention.cpp:117]: This led to duplication of tensor names in the model.
It copies the name of the "position_ids" to the unsqueeze tensor.

File: src/core/src/pass/sdpa_to_paged_attention.cpp (hunk 8 above)
Comment (itikhono) [reply src/core/src/pass/sdpa_to_paged_attention.cpp:89]: This led to duplication of tensor names in the model.
It copies the name of the "position_ids" to the unsqueeze tensor.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp [hunk 10]
```diff
@@ -94,6 +100,47 @@ ov::pass::TotalSequenceLengthPattern::TotalSequenceLengthPattern(
...
+    const std::shared_ptr<ov::op::v0::Parameter>& max_context_len) {
+    MATCHER_SCOPE(TotalSequenceLengthPatternQwen);
+
+    auto p_input_ids = wrap_type<v0::Parameter>();
+    auto p_unsqueeze = wrap_type<v0::Unsqueeze>({p_input_ids, any_input()});
+    auto p_opt_reshape_1 = optional<v1::Reshape>({p_unsqueeze, any_input()});
+    auto p_opt_convert_1 = optional<v0::Convert>(p_opt_reshape_1);
+    auto p_kv_shape_current = wrap_type<v3::ShapeOf>({p_opt_convert_1});
+    auto p_seq_current = wrap_type<v8::Gather>({p_kv_shape_current, any_input(), any_input()});
+    auto p_opt_convert_2 = optional<v0::Convert>(p_seq_current);
```
Comment (slyalin) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:117]: Can we check the symbol in a particular dimension in the shape of explicitly taken `input_id` input and code it as a predicate?

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp [hunk 11]
```diff
@@ -94,6 +100,47 @@ ov::pass::TotalSequenceLengthPattern::TotalSequenceLengthPattern(
...
+
+    auto p_max_context_len = wrap_type<v0::Parameter>();
+    auto p_prev_max_seq_len = wrap_type<v1::Subtract>({p_max_context_len, any_input()});
+    auto p_opt_convert_3 = optional<v0::Convert>(p_prev_max_seq_len);
+    auto p_opt_reshape_2 = optional<v1::Reshape>({p_opt_convert_3, any_input()});
```
Comment (slyalin) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:122]: Is this a pattern that we are generating in another part of SDPA2PA transformation? The original model doesn't have `max_context_len` parameter, so I think it is. If so, why do we need to match it? Can we re-arrange transformations somehow to keep `rt_info` that explicitly points to the desired node?

File: src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp (hunk 7 above)
Comment (itikhono) [reply src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp:537]: I used PrintModel transformation to generate the c++ code of the model.
But I had to split it by parts (attention mask, Q, K, V etc) manually. 
Not so sure if we have a way to automate it. Unfortunately, even the famous ov2py tool can't do it.
I'm planning to write the tests for the matcher passes as well so I believe the splitting was not so useless.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp [hunk 12]
```diff
@@ -94,6 +100,47 @@ ov::pass::TotalSequenceLengthPattern::TotalSequenceLengthPattern(
...
+
+        auto target_type = total_seq->get_output_element_type(0);
+        if (replacement->get_output_element_type(0) != target_type) {
+            replacement = std::make_shared<v0::Convert>(replacement, target_type);
+        }
+
+        auto required_shape = total_seq->get_output_partial_shape(0);
+        if (replacement->get_output_partial_shape(0) != required_shape && required_shape.rank().is_static()) {
+            replacement = op::util::reshapeTo(replacement, Shape(required_shape.rank().get_length(), 1));
+        }
```
Comment (slyalin) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:138]: That kind of shape/element_type adoption starts repeating. Could you move it into a single place?

File: src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp (hunk 6 above)
Comment (itikhono) [reply src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp:80]: Good catch!
I forgot to cut the weights :(

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp (hunk 5 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:35]: It was done by "replace" in the whole file.
I sincerely hope these few lines didn't get you in trouble with the review.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp (hunk 12 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:138]: Sure

File: src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp (hunk 3 above)
Comment (itikhono) [reply src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp:38]: Probably we don't need, I will double check

File: src/core/src/pass/sdpa_to_paged_attention.cpp (hunk 8 above)
Comment (slyalin) [reply src/core/src/pass/sdpa_to_paged_attention.cpp:89]: Well, it makes sense to copy the name because, in this particular place, we have two different interpretations of what we should treat as `position_ids`: the old one or the new one. Anyway, it doesn't trouble the user, because only input and output tensor names matter, everything inside is hidden. Why is it so important in this case?

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp (hunk 12 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp]: done

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp (hunk 4 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp]: done

File: src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp (hunk 3 above)
Comment (itikhono) [reply src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp]: done

File: src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp (hunk 2 above)
Comment (itikhono) [reply src/common/transformations/include/transformations/sdpa_to_paged_attention/position_ids_replacer.hpp]: done

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp (hunk 1 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/position_ids_replacer.cpp]: done

File: src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp (hunk 6 above)
Comment (itikhono) [reply src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp]: done

File: src/core/src/pass/sdpa_to_paged_attention.cpp (hunk 8 above)
Comment (itikhono) [reply src/core/src/pass/sdpa_to_paged_attention.cpp]: usually we do not allow to duplicate tensor names, I don't want to break this rule in this particular case

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp (hunk 10 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:125]: Not so sure this is a good solution for standalone transformation.
as I understand, we need to enable symbolic transformations for that.
I can discuss it later.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp (hunk 11 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:130]: I think pattern matching is a fairly robust way to detect this subgraph, and we have control on how we create this pattern in another transformation. 
I think we can leave it as is in this PR and consider it as a future improvement to this transformation.

Comment (jane-intel) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/total_sequence_length_pattern.cpp:130]: I think we need to run sdpa to pa transformations in symbolic environment, however, this should be a task for another PR.

**Review by jane-intel — APPROVED** [review]
